src/useRouteQueryParams/useRouteQueryParams.ts

```diff
@@ -0,0 +1,65 @@
+import { Ref } from 'vue'
+import { NoInfer } from '@/types/generics'
+import { isRouteParamClass, RouteParamClass } from '@/useRouteQueryParam/formats'
+import { useRouteQueryParam } from '@/useRouteQueryParam/useRouteQueryParam'
+
+type AnyRecord = Record<string, unknown>
+
+export type RouteQueryParamsSchema<T extends AnyRecord> = {
+  [P in keyof T]-?: NonNullable<T[P]> extends AnyRecord
+    ? RouteQueryParamsSchema<NonNullable<T[P]>>
+    : RouteParamClass<NonNullable<T[P]>>
+}
+
+export type RouteQueryParams<T extends AnyRecord> = {
+  [P in keyof T]-?: [T[P]] extends [AnyRecord | undefined]
+    ? [T[P]] extends [AnyRecord]
+      ? RouteQueryParams<T[P]>
+      : RouteQueryParams<Partial<T[P]>>
+    : Ref<T[P]>
+}
+
+export function useRouteQueryParams<T extends AnyRecord>(schema: RouteQueryParamsSchema<T>, defaultValue: NoInfer<T>): RouteQueryParams<T> {
+  return getSchemaRouteQueryParams(schema, defaultValue)
+}
+
+function isRouteParamSchema<T extends AnyRecord>(value: RouteQueryParamsSchema<T> | unknown): value is RouteQueryParamsSchema<T> {
+  return !isRouteParamClass(value)
+}
+
+function getSchemaRouteQueryParams<T extends AnyRecord>(
+  schema: RouteQueryParamsSchema<T>,
+  defaultValue: NoInfer<T>,
+  prefix?: string,
+): RouteQueryParams<T> {
+  const prefixed = (key: string): string => {
+    if (prefix) {
+      return `${prefix}.${key}`
+    }
+
+    return key
+  }
+
+  const params = Object.keys(schema).reduce<AnyRecord>((params, key) => {
+    const property = schema[key]
+    const propertyDefault = defaultValue[key]
+
+    if (isRouteParamClass(property)) {
+      params[key] = useRouteQueryParam(prefixed(key), property, propertyDefault)
+
+      return params
+    }
+
+    if (isRouteParamSchema(property)) {
```

Comment:
I'm guessing you did this to be highly explicit, but it seems you could replace this line with `else`, since this type guard just asserts that it's not `isRouteParamClass`

Reply:
If it wasn't for typescript I wouldn't need this if (or an else) at all. But anytime `Object.keys` is involved typescript forces you to be super explicit. I did try using an `else` but typescript doesn't really do well with type guards and else. It doesn't get the correct type and then I have a type error in the `getSchemaRouteQueryParams` call. 

Open to suggestions here. This was the only way I could figure out to appease typescript.

Reply:
hmmm bummer, the only alternative I can think of is casting the `property` on line 44. Don't think that's much better

Reply:
The thing that always gets me is typescript is technically correct. There could be other properties at run time (even though there aren't likely to be). 

Since this works and is pretty readable I'm going to leave it as is for now.